### PR TITLE
Changes to ensure if no object is opened user won't be able to write property name

### PR DESCRIPTION
**Pull request by shivam-sehgal**
### Purpose
Currently, even if the no object is opened or begin object is not called or all open objects are ended using jsonwriter.endObject() Still user is able to add json object property name, which shouldn't be happening
Closes #2407 

### Description
Currently when the Jsonwriter name method is called no check is there to identify if it's an empty document or all open objects are closed and stack size reaches 1 and the top element is NON_EMPTY_DOC basically the two case
* The user calls jsonwrite.name() for the first time in an empty document or when the beginobject method has not been called.
* The user calls jsonwrite.name() when all opened objects are closed.
In both of these invalid cases, the stack size should be 1 as per the current implementation 
In the first case, stack will be having one status only which will be EMPTY_DOCUMENT
 In the second case, all other statuses will be popped from the stack and only one status NONEMPTY_DOCUMENT will be present in the stack
So added a check in the name method for these 2 special cases
Also added and changed the unit test as per the requirement

### Checklist

- [ ] New code follows the [Google Java Style Guide](https://google.github.io/styleguide/javaguide.html)
- [ ] If necessary, new public API validates arguments, for example rejects `null`
- [ ] New public API has Javadoc
    - [ ] Javadoc uses `@since $next-version$`  
      (`$next-version$` is a special placeholder which is automatically replaced during release)
- [ ] If necessary, new unit tests have been added  
  - [ ] Assertions in unit tests use [Truth](https://truth.dev/), see existing tests
  - [ ] No JUnit 3 features are used (such as extending class `TestCase`)
  - [ ] If this pull request fixes a bug, a new test was added for a situation which failed previously and is now fixed
- [ ] `mvn clean verify javadoc:jar` passes without errors

### CODE REVIEW

File: gson/src/main/java/com/google/gson/stream/JsonWriter.java [hunk 1]
```diff
@@ -498,6 +498,10 @@ public JsonWriter name(String name) throws IOException {
     if (stackSize == 0) {
       throw new IllegalStateException("JsonWriter is closed.");
     }
+    // As currently in any valid case name function cannot be called if stack top has these 2 statuses
```
Comment (eamonnmcmanus) [inline gson/src/main/java/com/google/gson/stream/JsonWriter.java:501]: I found this comment a bit confusing. I think it isn't really necessary anyway? It doesn't really tell the reader anything that they can't already see from the following statement.

**Review by eamonnmcmanus — CHANGES_REQUESTED** [review]
Thanks, this looks great! I have some minor comments. I'm also going to try running this against all of Google's internal tests, though I don't expect that will find any problems.

File: gson/src/main/java/com/google/gson/stream/JsonWriter.java [hunk 2]
```diff
@@ -498,6 +498,10 @@ public JsonWriter name(String name) throws IOException {
...
+    if (stackSize == 1 && ( peek() == EMPTY_DOCUMENT || peek() == NONEMPTY_DOCUMENT )) {
```
[the remark this answers is not in the document]
Comment (shivam-sehgal) [reply gson/src/main/java/com/google/gson/stream/JsonWriter.java]: my bad, corrected that, thanks for pointing out

File: gson/src/test/java/com/google/gson/stream/JsonWriterTest.java [hunk 3]
```diff
@@ -114,13 +116,36 @@ public void testTopLevelValueTypes() throws IOException {
...
+    // Add an array of phone numbers (list of numbers)
+    List<Integer> phoneNumbers = Arrays.asList(1234567890, 98989, 9909);
+    jsonWriter.name("phoneNumbers").beginArray();
+    for (Integer phoneNumber : phoneNumbers) {
+      jsonWriter.value(phoneNumber);
     }
+    jsonWriter.endArray(); // End the array
+    jsonWriter.endObject(); // End the outer object
+    assertThrows(IllegalStateException.class, () -> jsonWriter.name("this_throw_exception_as_all_objects_are_closed"));
+    assertThrows(IllegalStateException.class, () -> jsonWriter.value("this_throw_exception_as_only_one_top_level_entry"));
```
[the remark this answers is not in the document]
Comment (shivam-sehgal) [reply gson/src/test/java/com/google/gson/stream/JsonWriterTest.java]: agree with you regarding status change and the possible impact of it, created a private method to return jsonwritter as you recommended, and running both tests with 2 separate objects

File: gson/src/test/java/com/google/gson/stream/JsonWriterTest.java [hunk 4]
```diff
@@ -114,13 +116,36 @@ public void testTopLevelValueTypes() throws IOException {
...
     StringWriter stringWriter = new StringWriter();
     JsonWriter jsonWriter = new JsonWriter(stringWriter);
-    jsonWriter.name("hello"); // TODO: This should throw, see https://github.com/google/gson/issues/2407
-    try {
-      jsonWriter.value("world");
-      fail();
-    } catch (IllegalStateException expected) {
-      assertThat(expected).hasMessageThat().isEqualTo("Nesting problem.");
+    assertThrows(IllegalStateException.class, () -> jsonWriter.name("hello"));
+    //removed nested exception test on jsonwriter.writevalue as it seems currently valid case per code
```
[the remark this answers is not in the document]
Comment (shivam-sehgal) [reply gson/src/test/java/com/google/gson/stream/JsonWriterTest.java]: makes sense, reading this comment will be odd for users as code is already gone, thanks for pointing it out, removed this comment

File: gson/src/main/java/com/google/gson/stream/JsonWriter.java (hunk 1 above)
Comment (shivam-sehgal) [reply gson/src/main/java/com/google/gson/stream/JsonWriter.java]: got it! removed this unecessary comment

**Comment by shivam-sehgal** [comment]
> 

Thank you for reviewing the code! I'll look forward to the test results.

**Review by eamonnmcmanus — APPROVED** [review]
Thanks for doing this! I ran it against all of Google's internal tests and did not see any problems.

**Comment by shivam-sehgal** [comment]
> Thanks for doing this! I ran it against all of Google's internal tests and did not see any problems.

Thanks for the approval :)

**Comment by Marcono1234** [comment]
Thanks @shivam-sehgal for your work on this!

Unfortunately I think my original issue made it look like this only affected the top level (which is addressed by your pull request), but it actually also affects trying to use `JsonWriter.name` inside a JSON array.

I have created #2476 to (hopefully) fully resolve this. It reverts some of your test changes because handling of multiple top-level values was already covered by multiple separate test methods, and while it is generally good that your method `getJsonWriterWithObjects` covered multiple value types; I think here more concise test data specific to this issue might be better. The data created by `getJsonWriterWithObjects` might rather be useful for integration tests / functional tests (I think) but there already a bunch of this in this repository. Sorry for that; I hope it is ok for you (and for @eamonnmcmanus).

**Pull request merged**